odxtools/udsbinner.py

```diff
@@ -0,0 +1,128 @@
+# SPDX-License-Identifier: MIT
+from io import StringIO
+from typing import Dict, Iterable, Optional
+
+from .diagservice import DiagService
+from .nameditemlist import NamedItemList
+from .parameters.codedconstparameter import CodedConstParameter
+
+
+class UdsBinner:
+    """Class to categorize a long list of services into the service
+    groups defined by the UDS standard.
+
+    This class is supposed to be used like this:
+
+    db = odxtools.load_file("my_cool_diagnostics_db.pdx")
+    ...
+    uds_bins = UdsBinner(db.ecus.my_ecu)
+    print(uds_bins)
+    """
+
+    def __init__(self, services: Iterable[DiagService]):
+        service_groups: Dict[Optional[int], NamedItemList[DiagService]] = {}
+        for service in services:
+            SID = self.__extract_sid(service)
+
+            if SID not in service_groups:
+                service_groups[SID] = NamedItemList()
+
+            service_groups[SID].append(service)
+
+        self.service_groups = service_groups
+
+    def __extract_sid(self, service: DiagService) -> Optional[int]:
+        # diagnostic services without requests are possible; just like
+        # aircraft without wings...
+        if service.request is None:
+            return None
+
+        prefix = 0  # prefix of constant parameters
+        cursor = 0  # bit position of the next parameter
+        for param in service.request.parameters:
+            if not isinstance(param, CodedConstParameter):
+                # we *need* at least the first byte of a request to be statically defined!
+                return None
+
+            param_len = param.get_static_bit_length()
+            if param_len is None:
+                return None
+
+            if not isinstance(param.coded_value, int):
+                return None
+
+            prefix <<= param_len
+            prefix |= param.coded_value & ((1 << param_len) - 1)
+            cursor += param_len
+
+            if cursor >= 8:
+                # we have a prefix that is at least 8 bits
+                # long. return its most significant byte.
+                prefix >>= cursor - 8
+                return prefix & 0xff
+
+        return None
+
+    def __str__(self) -> str:
+        """Return an informative string about which of the diagnostic
+        services are part of which UDS service group.
+
+        """
+        result = StringIO()
+        for sid, service_list in self.service_groups.items():
+            sid_name = self.sid_to_name.get(sid)
```

Comment:
will throw for unknown service,

```suggestion
            sid_name = self.sid_to_name.get(sid, hex(sid))
```

Reply:
I refactored this code with the goals of including the OBD services and using the existing `.uds` and `.obd` sub-packages...